### PR TITLE
Referrer policy test case

### PR DESCRIPTION
**Pull request by kristijanburnik**
Substantial refactoring of the test suite:
- A new simple API for generating and writing tests.
- Added initial test for no-refererrer - for now just basic assertions.

@mikewest, @jeisinger

### CODE REVIEW

**Comment by kristijanburnik** [comment]
## Update on the current state of the PR ##

* All the test logic now lies in /referrer-policy/generic/common.js and /referrer-policy/generic/referrer-policy-test-case.js

* Each test case is expanded (generated) from the specification in spec.json.

* The spec.json is currently reduced and not listing all the w3 spec requirements.

  * Some delivery methods and HTTP(S) protocol transitions are still omitted.

  * Only iframe and image is being tested as a subresource.

* Some tests fail - I'm investigating if it's due to non-existing implementation according to the new soon-to-be w3 spec for the test case or some other issue in my PR.

* There is an additonal test for decoding HTTP headers from a PNG image generated by ./referrer-policy/generic/image.py in ./referrer-policy/generic/image-decoding-test.html

## Testing the PR: ##

Once pulled in the branch: 

```
cd ~/web-platform-tests

# Generate all the test files.
./referrer-policy/generic/tools/generate.py

# Not sure if this is the right way, but ./manifest suggests the files are 
# generated from the source tree.
# Commit generated files for the MANIFEST.json
git checkout -b temp_referrer_policy
git add ./referrer-policy 
git commit -m "Temporary commit for testing Referrer Policy test cases."

# Generate the manifest and start the server.
./manifest && ./serve

# Navigate to http://web-platform.test:8000/tools/runner/index.html
# Run tests under path: "/referrer-policy"
# Click start
```

@mikewest, @jeisinger

**Comment by kristijanburnik** [comment]
This PR also resolves the URL port issue:
https://github.com/w3c/web-platform-tests/issues/1839

**Comment by mikewest** [comment]
Can you point me to the generated files as well? Perhaps they're in your repo somewhere? I'd like to get a feel for what's actually generated, but it's hard to do so just from the templates and python files.

Alternatively, you could update the README to help me generate the files myself. :)

**Comment by mikewest** [comment]
Nevermind, I needed to be in the right directory to execute `generate.py`.

**Comment by kristijanburnik** [comment]
I left some notes in the PR discussion. :-)

https://github.com/w3c/web-platform-tests/pull/1814



On Tue, May 19, 2015 at 6:43 PM, Mike West <notifications@github.com> wrote:

> Nevermind, I needed to be in the right directory to execute generate.py.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1814#issuecomment-103585510>
> .
>



-- 

*Kristijan Burnik*

Software Engineer Intern

burnik@google.com
+385 91 4846 379

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

Diese E-Mail ist vertraulich. Wenn Sie nicht der richtige Adressat sind,
leiten Sie diese bitte nicht weiter, informieren Sie den Absender und
löschen Sie die E-Mail und alle Anhänge. Vielen Dank.



This e-mail is confidential. If you are not the right addressee please do
not forward it, please inform the sender, and please erase this e-mail
including any attachments. Thanks.

**Comment by kristijanburnik** [comment]
I have now included tests for the referrer attribute on anchor and area elements.

The commit log is resquashed.

**Comment by kristijanburnik** [comment]
Added basic instructions and project overview to [README.md](https://github.com/kristijanburnik/web-platform-tests/blob/referrer-policy-test-case/referrer-policy/README.md)

**Comment by kristijanburnik** [comment]
Added tests for anchor element's rel="noreferrer". Area element currently excluded.

File: referrer-policy/spec.src.json [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+          "target_protocol": "*",
+          "delivery_method": ["http-csp", "meta-referrer", "meta-csp", "attr-referrer"],
+          "origin": "*",
+          "subresource": "*",
+          "referrer_url": "origin"
+        }
+      ]
+    },
+    {
+      "name": "origin-when-crossorigin",
```
Comment (mikewest) [inline referrer-policy/spec.src.json]: I typo'd this in the spec, and didn't think about it when implementing. Sorry for the confusion.

The actual policy is `origin-when-cross-origin`, and I've updated the spec accordingly. Chrome supports the correct string as of ~44, and I intend to deprecate and remove `origin-when-crossorigin` as soon as 44 rolls out to stable and GitHub/Dropbox switch to the correct string.

File: referrer-policy/generic/fetch.js [hunk 2]
```diff
@@ -0,0 +1,337 @@
+// Shamelessly copied from: https://github.com/github/fetch/blob/master/fetch.js
+// TODO(krisitijanburnik): See how to include third_party for WPT.
```
Comment (mikewest) [inline referrer-policy/generic/fetch.js]: I'm not sure it's a good idea to polyfill fetch in the tests. You'll be testing the polyfill rather than the feature in some subset of browsers. Did you consider simply PASSing the tests if `!window.fetch`?

Comment (mikewest) [inline referrer-policy/generic/fetch.js]: (Dropping the polyfill would also remove licensing concerns, etc.)

File: referrer-policy/get_failing_tests.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+
+import json
+
+data = json.load(open("attr-referrer-failing.json"))
+
+prefix = "http://web-platform.test:8000"
```
Comment (mikewest) [inline referrer-policy/get_failing_tests.py]: This only works locally. Is this file intended to do anything when requested over the web? If not, I'd suggest moving it to tools, and adding some comments about its intended use case.

File: referrer-policy/generic/template/test.html.template [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      method and when the target request is %(origin)s.</p>
+
+    <p>See <a href="%(spec_specification_url)s" target="_blank">specification</a>
+       details for this test.</p>
+
+    <h3>Scenario outline</h3>
+    <table>
+      <tbody>
+        <tr>
+          <th>Delivery method</th>
```
Comment (mikewest) [inline referrer-policy/generic/template/test.html.template]: Tiny nit: I'd have written this as:

```
<dl>
  <dt>Delivery Method</dt>
  <dd>%(delivery_method)s</dd>
  ...
</dl>
```

File: referrer-policy/generic/template/test.html.template [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    <script src="/referrer-policy/generic/fetch.js"></script>
+    <!-- Common global functions for referrer-policy tests. -->
+    <script src="/referrer-policy/generic/common.js"></script>
+    <!-- The original specification JSON for validating the scenario. -->
+    <script src="/referrer-policy/spec_json.js"></script>
+    <!-- Simple wrapper API for all referrer-policy test cases. -->
+    <script src="/referrer-policy/generic/referrer-policy-test-case.js?pipe=sub"></script>
+  </head>
+  <body>
+    <h1>%(spec_title)s</h1>
```
Comment (mikewest) [inline referrer-policy/generic/template/test.html.template]: The text is going to cause problems for automated tests. Blink's `testharness.js` support scans the resulting file, and dies horribly if lines show up that don't start with `PASS`. Not a reason to delay landing this patch, but it's something we're going to have to refactor before we can import these tests cleanly.

File: referrer-policy/generic/subresource/xhr.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import json
```
Comment (mikewest) [inline referrer-policy/generic/subresource/xhr.py]: Nit: I'd have named this file something like `echo-headers.py`. It's equally useful for XHR. and `fetch()`.

File: referrer-policy/generic/subresource/image.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    response.add_required_headers = False
+    response.writer.write_status(200)
+    # Allow cross-origin access to get pixel data via JS.
+    response.writer.write_header("access-control-allow-origin", "*")
+    response.writer.write_header("content-type", "image/png")
+    response.writer.write_header("cache-control", "no-cache; must-revalidate")
+    response.writer.end_headers()
+
+    headers_as_json = json.dumps(request.headers)
+    body = encode_string_as_png_image(str(headers_as_json))
```
Comment (mikewest) [inline referrer-policy/generic/subresource/image.py]: This is totally clever. I love it. :)

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      assert_in_array(scenario[field], expectedFields[field],
+                      "Scenario's " + field + " is one of: " +
+                      expectedFields[field].join(", ")) + "."
+    }
+
+    // Check if the protocol is matched.
+    assert_equals(scenario["source_protocol"] + ":", location.protocol,
+                  "Protocol of the test page should match the scenario.")
+
+  }, "[ReferrerPolicyTestCase] The test scenario is valid.");
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: I don't think it's worth doing this on every run; it makes sense to verify that we've generated sane code once, but it doesn't make sense to pay that cost in processing time every time we run any test.

I'd suggest either removing it right now, or adding a `TODO(burnik)` to kill it once you're confident that the generated code is correct.

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+  var subresourceInvoker = {
+    "a-tag": queryLink,
+    "area-tag": queryAreaLink,
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: I think we can drop `<area>`, honestly. It should go through the same paths as `<a>`, and there's a fairly significant cost in terms of runtime for each tag we add. I think `<a>`, ``, `<script>`, and `<iframe>` is a reasonable sample to concentrate on.

If you'd like to keep `<area>`, then I think we need to go in the other direction, and cover _every_ mechanism by which a resource can be loaded. That doesn't seem like a good use of your time, but I'm happy to let you spend it there if you like. :)

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    "fetch-request": queryFetch,
+    "iframe-tag": queryIframe,
+    "img-tag":  queryImage,
+    "script-tag": queryScript,
+    "worker-request": queryWorker,
+    "xhr-request": queryXhr
+  };
+
+  var pathForSubresource = SPEC_JSON["subresource_path"];
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: I'd suggest hard-coding this parameter into the generated `scenario` rather than reading the `spec_json.js` file in every test.

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+  var referrerUrlResolver = {
+    "omitted": function() {
+      return undefined;
+    },
+    "origin": function() {
+      return location.protocol + "//" + location.hostname +
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: Why not `document.origin`?

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+             normalizePort(location.port) + "/";
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: Origins usually don't have trailing slashes. Are we sending a trailing slash in the header?

Hrm. It looks like we do. Why do we do that? I know I wrote the code and the spec, but I disagree with myself. :) We don't add a `/` for the `Origin:` header, so we shouldn't add it here either.

Let's discuss in https://github.com/w3c/webappsec/issues/382 and come back to this later.

File: referrer-policy/generic/template/test.html.template (hunk 5 above)
Comment (mikewest) [inline referrer-policy/generic/template/test.html.template]: See http://testthewebforward.org/docs/test-templates.html#script-test-with-metadata for the recommended test format.

File: referrer-policy/generic/referrer-policy-test-case.js [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  };
+
+  var t = {
+    _scenario: scenario,
+    _testDescription: testDescription,
+    _subresourceUrl: null,
+    _expectedReferrerUrl: null,
+    _constructSubresourceUrl: function() {
+      var domainForOrigin = {
+        "cross-origin": "www1." + location.hostname,
```
Comment (mikewest) [inline referrer-policy/generic/referrer-policy-test-case.js]: http://wptserve.readthedocs.org/en/latest/pipes.html#sub gives you the ability to pull hosts via the `{{...}}` method (which you used below for ports). Why not use it here?

**Comment by mikewest** [comment]
This is looking amazing. Comments inline.

File: referrer-policy/get_failing_tests.py (hunk 3 above)
Comment (kristijanburnik) [inline referrer-policy/get_failing_tests.py]: Yes, sry about this one, it slipped through (because of the manifest mechanism, which is quite bad for development), I've used it to debug my Chromium CL and it helped. :)

File: referrer-policy/generic/template/test.html.template (hunk 4 above)
Comment (kristijanburnik) [inline referrer-policy/generic/template/test.html.template]: I'll consider it if we decide to keep this meta. :-)

File: referrer-policy/generic/template/test.html.template (hunk 5 above)
Comment (kristijanburnik) [inline referrer-policy/generic/template/test.html.template]: Added a TODO into the template.

Actually I was thinking: I could add an additional template matching [this format](http://testthewebforward.org/docs/test-templates.html#script-test-with-metadata) and extending the ```./generic/tools/generate.py``` with a command line flag to choose which template to use. 

We can then generate the tests for other purposes as well.

File: referrer-policy/generic/subresource/xhr.py (hunk 6 above)
Comment (kristijanburnik) [inline referrer-policy/generic/subresource/xhr.py]: This file is already being used for both:
```json
 "subresource_path": {   
   ...
   "fetch-request": "/referrer-policy/generic/subresource/xhr.py",
   ...
   "xhr-request": "/referrer-policy/generic/subresource/xhr.py"
  }
```

Let's consider the subresources directory and naming:

```
subresource
├── document.py
├── image.py
├── script.py
├── worker.py
└── xhr.py
```

Against:

```
subresource
├── document.py
├── image.py
├── script.py
├── worker.py
└── echo-headers.py
```

Fetch is still very associative with XHR so that's my reasoning. :-)
Other scripts also echo the headers, just in a slightly different fashion.

File: referrer-policy/generic/subresource/image.py (hunk 7 above)
Comment (kristijanburnik) [inline referrer-policy/generic/subresource/image.py]: I'm glad you approve. :-) I've even considered using RGBA, but it got messy quite quickly so I'm sticking to RGB for the time being.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 8 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: I think I can get multiple birds with one stone:

Supplying a specific template when generating, e.g. ```./generic/tools/generate.py --target=debug``` or ```./generic/tools/generate.py --target=release```,
would help me move these sanity checks outside of the test case (into a ```sanity-checks.js``` just for debugging) and reduce the overall resources for each test when in production. What do you think? :-)

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 9 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: This actually lists all supported sub-resources and their handlers, however some of them get suppressed by the ```spec.src.json```. I would keep them here for now to still have support for extending when needed, but I might refactor this a bit.

With the spec JSON we can reduce the number of different sub-resources we test for in particular circumstances. See more on the ```excluded_tests``` section.

I am also concerned with the performance of the test suite as it currently runs for ~ 4 minutes on my machine. There will definitely be tests which we should omit.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 10 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: Agreed. This made sense for reusability since I'm already using the ```spec_json.js``` for validating the test itself.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 11 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: Done.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 12 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: Ok. I've also added a comment on this topic.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 13 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: Good point. Done. Also added a TODO note.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 11 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: I totally missed out on that property. Done.

File: referrer-policy/generic/referrer-policy-test-case.js (hunk 8 above)
Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: Agreed. Adding a debug/release template and a JS file for sanity checks should resolve this.

Comment (kristijanburnik) [inline referrer-policy/generic/referrer-policy-test-case.js]: \* JS file for sanity checks in debug mode only.

File: referrer-policy/spec.src.json (hunk 1 above)
Comment (kristijanburnik) [inline referrer-policy/spec.src.json]: Yes, I've seen that CL and am aware of the issue regarding these tests as well.
I'll add this to my personal TODO list. 
I'll also see what I can do to allow adding comments to the spec JSON.

File: referrer-policy/generic/fetch.js (hunk 2 above)
Comment (kristijanburnik) [inline referrer-policy/generic/fetch.js]: Done. Passing tests for browsers which don't support fetch. However, there is a subresource-test which will fail on those.

**Comment by kristijanburnik** [comment]
Thanks! Addressed nits and re-squashed.

**Comment by mikewest** [comment]
This looks pretty good. Can you please fix the (hopefully trivial) linting error introduced by this patch ("TRAILING WHITESPACE: referrer-policy/README.md line 3"; you can run 'lint' locally to reproduce), then generate the tests, commit them, and re-upload?

Please _don't_ squash the generation step into 4b82e0f. Otherwise the diff will be impossible for folks to understand. :)

**Comment by kristijanburnik** [comment]
Done!

Also:
- suppressing area-tag for now
- failing Fetch sub-resource test early not to block the runner
- fixed lint issues in generated files

Stats:
497 tests + 8 sanity checkers (subresource-test).

Approx. test suite runtime (64bit | 12 cores | 32 GiB RAM | Ubuntu 14.04):
~ 1.5 minutes on chrome stable (v. 42)
~ 1.5 minutes on firefox stable (v. 37.0.2)
~ 4 minutes on a chromium debug build

Size of repo increased by ~ 5.3M (2.7%)
This increase will be reduced to nearly a half when compacted the templates.

GitHub is not happy while displaying a diff of 662 files. :-/



On Thu, May 21, 2015 at 2:58 PM, Mike West <notifications@github.com> wrote:

> This looks pretty good. Can you please fix the (hopefully trivial) linting
> error introduced by this patch ("TRAILING WHITESPACE:
> referrer-policy/README.md line 3"; you can run 'lint' locally to
> reproduce), then generate the tests, commit them, and re-upload?
>
> Please *don't* squash the generation step into 4b82e0f
> <https://github.com/w3c/web-platform-tests/commit/4b82e0f1fe175b576ea8a841561727a21bd7c0e2>.
> Otherwise the diff will be impossible for folks to understand. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/1814#issuecomment-104263743>
> .
>



-- 

*Kristijan Burnik*

Software Engineer Intern

burnik@google.com
+385 91 4846 379

Google Germany GmbH

Dienerstraße 12

80331 München

Geschäftsführer: Graham Law, Christine Elizabeth Flores

Registergericht und -nummer: Hamburg, HRB 86891

Sitz der Gesellschaft: Hamburg

Diese E-Mail ist vertraulich. Wenn Sie nicht der richtige Adressat sind,
leiten Sie diese bitte nicht weiter, informieren Sie den Absender und
löschen Sie die E-Mail und alle Anhänge. Vielen Dank.



This e-mail is confidential. If you are not the right addressee please do
not forward it, please inform the sender, and please erase this e-mail
including any attachments. Thanks.

**Comment by mikewest** [comment]
Ok. I'll land this, and we can chat today about followup patches. In particular, I think it's important to strip the test templates down to remove some of the debug context that you've provided; that's going to be necessary if we ever want to run these tests as part of Blink's test suite (and I believe Firefox and WebKit have similar requirements).

**Comment by kristijanburnik** [comment]
I have sent out an e-mail about this new test suite to the public-webappsec@
list.
I'm just enlisting some of the work that's planned from my end.

Here's some of the planned maintenance that's currently in development for
the Referrer Policy test suite:
- reduce size of the HTML test file and strive to increase run-time
performance
- extend the test generator to support command line flags for selecting
templates
- remove sanity checks when in production
- rearrange the test suite into debug/release mode for getting best out of
both worlds
- determine which tests are exhibiting duplicate behavior and/or can be
excluded
- HTTP redirection tests
- HTTPS to HTTP downgrades

P.S.
And by WebKit you mean Safari? :-)

**Comment by jgraham** [comment]
Can I just remind folks that it's *not* OK to merge PRs that fail the lint check. Please either fix the error or, if necessary, update the whitelist. Don't leave master in a failing state because it stops the lint being useful for other PRs.

If there are specific rules you disagree with, let's take it up on the mailing list.

**Comment by kristijanburnik** [comment]
The lint issue is not related to this PR. 

I believe this is the cause:
https://github.com/w3c/web-platform-tests/commit/38a1de37d38118d4c841ff5bd494aaa350329f9b#diff-dfe318f5fcb259e1557a1fc898b1d944

I've already mentioned this to the author.

**Comment by mikewest** [comment]
@jgraham: I filed https://github.com/w3c/web-platform-tests/pull/1847 to fix the linting script.

Note that, as @kristijanburnik noted, the breakage has been ongoing for ~11 days, and had nothing to do with this patch. (Note also that @kristijanburnik _did_, in fact, fix a linting error an earlier version of this patch introduced (https://github.com/w3c/web-platform-tests/pull/1814#issuecomment-104263743); he's totally doing the right thing here. :) )
